### PR TITLE
Metadata validation

### PR DESCRIPTION
**Pull request by theo-m**
- `pydantic` metadata schema with dedicated validators against our taxonomy
- ci script to validate new changes against this schema and start a vertuous loop
- soft validation on tasks ids since we expect the taxonomy to undergo some changes in the near future

for reference with the current validation we have ~365~ 378 datasets with invalid metadata! full error report [_here_.](https://gist.github.com/theo-m/61b3c0c47fc6121d08d3174bd4c2a26b)

### CODE REVIEW

File: src/datasets/utils/resources/creators.json
Comment (theo-m) [inline src/datasets/utils/resources/creators.json:1]: these files were brought from https://github.com/huggingface/datasets-tagging

File: src/datasets/utils/metadata.py
Comment (lhoestq) [inline src/datasets/utils/metadata.py:12]: We should add `importlib_resources` to the test dependencies for python<3.7 then ?

This makes me also think we should have a CI for python 3.6

File: src/datasets/utils/resources/creators.json
Comment (lhoestq) [reply src/datasets/utils/resources/creators.json:1]: Is it possible to have these json files in a `resources` folder at the root of the repo or something like that ?
So that it's easier to find them

Is that compatible with how importlib.resources works ? Or do they have to be inside the `src/datasets` directory ?

File: src/datasets/utils/metadata.py
Comment (lhoestq) [inline src/datasets/utils/metadata.py:53]: would be nice to have json files for those as well

File: src/datasets/utils/metadata.py
Comment (lhoestq) [inline src/datasets/utils/metadata.py:143]: you can ignore any tag value containing "other" (not only at the beginning of the tag value)

For example: `text-classification-other-pokemon-classification`

(no, there's no dataset with this tag (yet))

**Review by lhoestq — COMMENTED** [review]
This is really cool thanks :)

I left a few comments.

Also I was wondering this is really needed to have `utils.metadata` as a submodule of `datasets` ? This is only used by the CI so I'm not sure we should have this in the actual `datasets` package.

File: src/datasets/utils/metadata.py
Comment (theo-m) [reply src/datasets/utils/metadata.py:12]: tbh I added this copy pasting from SO, I didn't look at the repo's policy for python versions support.

File: src/datasets/utils/metadata.py
Comment (theo-m) [reply src/datasets/utils/metadata.py:143]: that's true for `tasks_ids` but not categories

File: src/datasets/utils/resources/creators.json
Comment (theo-m) [reply src/datasets/utils/resources/creators.json:1]: I find that it being in the same directory as the schema definition is better, easier to find for lib devs and for the others the url to the directory would be enough. 
wdyt?

**Comment by theo-m** [comment]
> Also I was wondering this is really needed to have `utils.metadata` as a submodule of `datasets` ? This is only used by the CI so I'm not sure we should have this in the actual `datasets` package.

I'm unclear on the suggestion, would you rather have a root-level `./metadata.py` file? I think it's well where it is, if anything we could move it out of utils and into `datasets` as it could be used by e.g. `DatasetDict` so that users can pull the metadata easily rather than have to reparse the readme.

**Comment by lhoestq** [comment]
Ok that makes sense if we want to have functions that parse the metadata for users

File: src/datasets/utils/metadata.py
Comment (lhoestq) [reply src/datasets/utils/metadata.py:12]: You just need to add
```python
"importlib_resources;python_version<'3.7'"
```
to the metadata validation requirements:
```python
    # metadata validation
    "langcodes[data]>=3.1.0",
    "pydantic>=1.8.1",
    "importlib_resources;python_version<'3.7'",
```

Comment (theo-m) [reply src/datasets/utils/metadata.py:12]: ooh i understand now, i thought it was already in the stdlib but under a different name!

**Comment by gchhablani** [comment]
Hi @theo-m @lhoestq 

This seems very interesting. Should I add the descriptions to the PR on `datasets-tagging`? Alternatively, I can also create a google-sheet/markdown table :)

Sorry for the delay in responding.

Thanks,
Gunjan

**Comment by theo-m** [comment]
> Hi @theo-m @lhoestq
> 
> This seems very interesting. Should I add the descriptions to the PR on `datasets-tagging`? Alternatively, I can also create a google-sheet/markdown table :)
> 
> Sorry for the delay in responding.
> 
> Thanks,
> Gunjan

Hi @gchhablani, yes I think at the moment the best solution is for you to write in `datasets-tagging`, as the PR will allow us to discuss and review, even though the work will be ported to this repo in the end. 
Or we wait for this to be merged and you reopen the PR here, your call :)

File: src/datasets/utils/metadata.py
Comment (theo-m) [reply src/datasets/utils/metadata.py:172]: up to you, the idea was to keep a simple script that one could run to validate a readme in using `python -m datasets.utils.metadata myreadme.md`

Comment (lhoestq) [reply src/datasets/utils/metadata.py:172]: We usually add scripts to the `datasets-cli`. This would be a nice addition though.
We can keep this here if it's considered experimental.

Comment (lhoestq) [reply src/datasets/utils/metadata.py:172]: let's keep it here for now then

**Comment by SBrandeis** [comment]
cc @abhi1thakur

File: src/datasets/utils/resources/languages.json
Comment (lhoestq) [inline src/datasets/utils/resources/languages.json:1]: Where did you get this data from ? (for future references and eventual updates)

**Review by lhoestq — APPROVED** [review]
Thank you @theo-m @SBrandeis !

File: src/datasets/utils/resources/languages.json
Comment (SBrandeis) [reply src/datasets/utils/resources/languages.json:1]: I got the language codes from `datahub.io` (https://datahub.io/core/language-codes) and mapped the Language names manually with the `langcodes` package.
I have added a comment here: https://github.com/huggingface/datasets/pull/2107/files#diff-00db9a5082ea37ce6236856b9c01bdcca69facd06a0d9484141dd2c81c62f5f0R30-R32
Let me know if the statement is not clear 😄

Comment (lhoestq) [reply src/datasets/utils/resources/languages.json:1]: Ok thanks :)

**Pull request merged**
